### PR TITLE
WIP: normalize tool output to BED format

### PR DESCRIPTION
**Pull request by brentp**
each tool currently outputs its own format. This PR is to make them all output BED. If users request `--columns chrom, start, end`, then those are ignored since chrom, start, end are always prefixed to be the first columns reported.

See changes in test-de-novo.sh . @arq5x does this seem the best way to go?

Todo
+ [ ] adjust non inheritance tools (compound_het, etc)
+ [ ] update tests (only de_novo updated so far).
+ [ ] re-open this PR on dev branch once it is up-to-date.

### CODE REVIEW

**Comment by bgruening** [comment]
Great idea!  @brentp thanks for unifying the output!

**Comment by brentp** [comment]
@bgruening , glad it will be useful. Other than outputting BED (prefixing the current output with chrom, start, end), is there any other standardization that will make downstream use simpler?

**Comment by dgaston** [comment]
In general the more similar the output of all those tools is to the standard gemini query format the better.

**Comment by bgruening** [comment]
@brentp sticking to standards is all what I can imagine/wish (in general, actually gemini do a good job in this already.)
Does the switch to turn off the header will stay?

**Comment by brentp** [comment]
@bgruening this PR doesn't affect the main query program, just tools like de_novo and autosomal_recessive which don't have a switch to turn off the header. Is that what you mean?

@dgaston , does the output in the tests (https://github.com/arq5x/gemini/pull/436/files#diff-6467e2e50520e0dee563c76935e74e05R15) as changed seem reasonable. This is different from the main query format in that certain output columns *must* be reported.

**Comment by bgruening** [comment]
@brentp in Galaxy such a switch would be handy, otherwise I can simple drop the first line, both works for me.

**Comment by dgaston** [comment]
@brentp So for instance, why not just have the genotype columns reported as they are currently in the query format (each individual has their own column). The additional column of the family member details is ok.

**Comment by brentp** [comment]
@dgaston, that makes sense to me. So only required extra columns of: 

+ family_id
+ family_members (comma delimited in single column)
+ family_genotypes (comma delimited)

and other columns can be requested with gt.* ? Those 3 columns will always appear at the end of the output. Then users can request --columns "chrom, start, end, ..." to get BED formatted output (this would remove the automatic bed output in commit above).

I'll wait for others to weigh in.

**Comment by arq5x** [comment]
Summing up, we are proposing:

1. If `--columns` is not used, _all_ columns will be reported, followed by family_id, family_members, and family_genotypes.
2. If `--columns` is  used, those columns will be reported, followed by family_id, family_members, and family_genotypes.
3. If BED format is wanted, users should start the `--columns` as : `--columns "chrom,,start,end"`

The compound heterozygotes tool is a special case because it has an extra comp_het_id column. In its case, the comp_het_id will follow (or precede?) the above four columns.

Also, the forthcomiung Mendelian violations tool will adhere to these format guidlines.

@jxchong - Given your feedback, I am wondering if you have any comments about these proposed changes?

**Comment by jxchong** [comment]
I think these changes sound good.  I think for situation #1, the default should be all columns, with the columns beginning with chrom, start, end.

I think the comp_het tool should follow the BED format columns and have comp_het_id follow after (could just be column 4/5 or the last column?)

On April 13, 2015 at 12:16:07 PM, Aaron Quinlan (notifications@github.com) wrote:

Summing up, we are proposing:

If --columns is not used, all columns will be reported, followed by family_id, family_members, and family_genotypes.
If --columns is used, those columns will be reported, followed by family_id, family_members, and family_genotypes.
If BED format is wanted, users should start the --columns as : --columns "chrom,,start,end"
The compound heterozygotes tool is a special case because it has an extra comp_het_id column. In its case, the comp_het_id will follow (or precede?) the above four columns.

Also, the forthcomiung Mendelian violations tool will adhere to these format guidlines.

@jxchong - Given your feedback, I am wondering if you have any comments about these proposed changes?

**Comment by brentp** [comment]
I have made the update to the change so that it is as discussed. If they don't request anything, then we just ask for "*" which will output chrom, start, end first since those are the first 3 columns in the database.

For the coming mendelian error tool, it will output requested columns along with a "violation" column indicating the type of violation (LOH/disomy, de novo, etc.). I will likely add a probability column that uses the PL's if available for mendelian errors too.

Adding tests and working on comp_het tool.

**Comment by arq5x** [comment]
@brentp and I discussed this a bit more offline. 

1.  The `sample` column. There is an additional complexity for the `comp_hets` tool in that, unlike the others, it also adds a `sample` column reflecting the sample in which the CH was observed. I actually think this is a useful default column for the built in disease model tools, as simply listing family ids and genotypes makes it sometimes difficult to see which samples (particularly in a large family with multiple offspring) had the variant/event.  For example, if there are two or more affected individuals in a family, the report a de novo mutation would be slightly difficult to see by eye (or inspect programiatically) if when just one of the two kids has the de novo.  A `sample` column would state which of the two had the event.  For other disease models, the `sample` column could be used to list the samples that had the event.
*Therefore, I propose that all tools report a `samples` column after the aforementioned `family_id`, `family_members`, and `family_genotypes` columns.*

2. A common `variant_id` column. The `comp_hets` tool also uses a `comp_het_id` to track which lines represent a given comp het candidate (i.e., two rows per event, inherently). In the interest of being able to concatenate the results of multiple tools (e.g., running both `autosomal_recessive` and `comp_hets` in search of recessive variants), I propose that all tools report a unified `variant_id` column reflecting the `variant_id` column the variants table.  It should come before the `family_id`, `family_members`, `family_genotypes`, and `samples` columns. For all tools other than `comp_hets`, it should just be the `variant_id` in the table. For `comp_hets`, it should be a composite: `variant_id` + `_` + `comp_het_id`.  For example:  if the first comp_het candidate was made up variant 7 and 9, the variant_ids for the two rows would be 1_7 and 1_9.

In summary, all tools should have the following fixed output columns that will always follow any columns specified by `--columns`:

- variant_id
- family_id
- family_members (comma delimited in single column)
- family_genotypes (comma delimited)
- samples

So, if the user specified `--columns "chrom,start,end"`, the output would be:

- chrom
- start
- end
- variant_id
- family_id
- family_members (comma delimited in single column)
- family_genotypes (comma delimited)
- samples

We are interested in your further thoughts, as we would like to wrap this up for the next release.

**Comment by jxchong** [comment]
I think these additions are a GREAT idea and will make it considerably easier to combine analyses across multiple tools.

How would you represent variants with multiple families (and family members) with the same mutation? You’d have multiple family_ids (easy) but then possibly multiple family_members, family_genotypes, and samples per family_id,


On April 13, 2015 at 5:22:44 PM, Aaron Quinlan (notifications@github.com) wrote:

@brentp and I discussed this a bit more offline.

The sample column. There is an additional complexity for the comp_hets tool in that, unlike the others, it also adds a sample column reflecting the sample in which the CH was observed. I actually think this is a useful default column for the built in disease model tools, as simply listing family ids and genotypes makes it sometimes difficult to see which samples (particularly in a large family with multiple offspring) had the variant/event. For example, if there are two or more affected individuals in a family, the report a de novo mutation would be slightly difficult to see by eye (or inspect programiatically) if when just one of the two kids has the de novo. A sample column would state which of the two had the event. For other disease models, the sample column could be used to list the samples that had the event.
Therefore, I propose that all tools report a samples column after the aforementioned family_id, family_members, and family_genotypes columns.

A common variant_id column. The comp_hets tool also uses a comp_het_id to track which lines represent a given comp het candidate (i.e., two rows per event, inherently). In the interest of being able to concatenate the results of multiple tools (e.g., running both autosomal_recessive and comp_hets in search of recessive variants), I propose that all tools report a unified variant_id column reflecting the variant_id column the variants table. It should come before the family_id, family_members, family_genotypes, and samples columns. For all tools other than comp_hets, it should just be the variant_id in the table. For comp_hets, it should be a composite: variant_id + _ + comp_het_id. For example: if the first comp_het candidate was made up variant 7 a nd 9, th e variant_ids for the two rows would be 1_7 and 1_9.

In summary, all tools should have the following fixed output columns that will always follow any columns specified by --columns:

variant_id
family_id
family_members (comma delimited in single column)
family_genotypes (comma delimited)
samples
So, if the user specified --columns "chrom,start,end", the output would be:

chrom
start
end
variant_id
family_id
family_members (comma delimited in single column)
family_genotypes (comma delimited)
samples
We are interested in your further thoughts, as we would like to wrap this up for the next release.

**Comment by arq5x** [comment]
Good question, @jxchong.

I see two ways of handling that.

1. Use a different delimiter (e.g., "|" or ";") to separate the comma-separated `family_id` and `family_genotype` lists.
2. Report each family on a separate line and use the common `variant_id` to track the fact that the same variant was observed in multiple families.

From a output cleanliness perspective, I prefer #2. But I suspect #1 will be more useful from a data interpretation perspective.

**Comment by brentp** [comment]
I like 2. It's easier to parse and avoids packing stuff into an already overridden column.

**Comment by arq5x** [comment]
I prefer 2 as well.  The case where it gets a bit harder is `comp_hets` where the `variant_id` is a composite of `variant_id` and `comp_het_id`.  But I agree that it is cleaner for sure. 

One thing that could be done to further facilitate parsing and downstream filtering is to add one more column to the default set of columns called `family_count`, which would reflect the count of families that were observed to have a given variant id.  This would trigger the researcher to know that even though the given line reflects one family, there were N-1 other families with the same variant.

**Comment by jxchong** [comment]
I like adding family_count as a column to make it easier to visually parse.



On April 13, 2015 at 6:01 PM, Aaron Quinlan (notifications@github.com) wrote:

I prefer 2 as well. The case where it gets a bit harder is comp_hets where the variant_id is a composite of variant_id and comp_het_id. But I agree that it is cleaner for sure.

One thing that could be done to further facilitate parsing and downstream filtering is to add one more column to the default set of columns called family_count, which would reflect the count of families that were observed to have a given variant id. This would trigger the researcher to know that even though the given line reflects one family, there were N-1 other families with the same variant.

**Comment by brentp** [comment]
One potential source of confusion is that the `--min-kindreds` argument operates by gene. This will now  introducing a new, but similar count that is by variant.

**Comment by arq5x** [comment]
Hmm, good point. However, the total `family_count` across variants will always be >= the `--min-kindred` setting. That said, I agree it might be a bit confusing.

**Comment by jxchong** [comment]
Just throwing out another idea: just report the list of family_ids without reporting the family members with the variant?  In theory since this is (only?) an issue when using the built-in analysis tools, you should already be able to infer the genotypes of the family members for that variant since you already know the variant is in that family?


On April 14, 2015 at 10:33:07 AM, Aaron Quinlan (notifications@github.com) wrote:

Hmm, good point. However, the total family_count across variants will always be >= the --min-kindred setting. That said, I agree it might be a bit confusing.

**Comment by brentp** [comment]
Here is the output from a few tools. I believe it matches what was proposed. Please let me know any changes:

mendel
======

```
brentp@x1:/usr/local/src/gemini/test$ gemini mendel_errors  --columns "chrom, start, end, ref, alt, impact, impact_severity" test.auto_dom.no_parents.db | cols
WARNING: Unable to find parents for family (2). GEMINI is currently only able to identify candidates from two generational families.
chrom  start     end       ref  alt  impact          impact_severity  violation  variant_id         family_id  family_members                                                     family_genotypes  affected_samples  family_count
chr10  1142207   1142208   T    C    stop_loss       HIGH             1          plausible de novo  1          1_dad(dad;unaffected),1_mom(mom;unaffected),1_kid(child;affected)  T/T,T/T,T/C       1_kid             1
chr10  48003991  48003992  C    T    non_syn_coding  MED              3          plausible de novo  1          1_dad(dad;unaffected),1_mom(mom;unaffected),1_kid(child;affected)  C/C,C/C,C/T       1_kid             1
chr10  48004991  48004992  C    T    non_syn_coding  MED              4          plausible de novo  1          1_dad(dad;unaffected),1_mom(mom;unaffected),1_kid(child;affected)  C/C,C/C,C/T       1_kid             1
```

autosomal dominant
===============

```
brentp@x1:/usr/local/src/gemini/test$ gemini autosomal_dominant  --columns "chrom, start, end, ref, alt, impact_severity" test.auto_dom.db | cols
chrom  start      end        ref  alt  impact_severity  gene    variant_id  family_id  family_members                                                   family_genotypes  affected_samples  family_count
chr10  48003991   48003992   C    T    MED              ASAH2C  3           3          3_dad(dad;affected),3_mom(mom;unknown),3_kid(child;affected)     C/T,C/C,C/T       3_dad,3_kid       2
chr10  48003991   48003992   C    T    MED              ASAH2C  3           2          2_dad(dad;unaffected),2_mom(mom;affected),2_kid(child;affected)  C/C,C/T,C/T       2_mom,2_kid       2
chr10  48004991   48004992   C    T    MED              ASAH2C  4           3          3_dad(dad;affected),3_mom(mom;unknown),3_kid(child;affected)     C/T,C/C,C/T       3_dad,3_kid       2
chr10  48004991   48004992   C    T    MED              ASAH2C  4           2          2_dad(dad;unaffected),2_mom(mom;affected),2_kid(child;affected)  C/C,C/T,C/T       2_mom,2_kid       2
chr10  135336655  135336656  G    A    LOW              SPRN    5           3          3_dad(dad;affected),3_mom(mom;unknown),3_kid(child;affected)     G/A,G/G,G/A       3_dad,3_kid       1
chr10  1142207    1142208    T    C    HIGH             WDR37   1           3          3_dad(dad;affected),3_mom(mom;unknown),3_kid(child;affected)     T/C,T/T,T/C       3_dad,3_kid       2
chr10  1142207    1142208    T    C    HIGH             WDR37   1           2          2_dad(dad;unaffected),2_mom(mom;affected),2_kid(child;affected)  T/T,T/C,T/C       2_mom,2_kid       2
```

comp_het
========

```
brentp@x1:/usr/local/src/gemini/test$ gemini comp_hets     --ignore-phasing     --columns "chrom, start, end"     --families 3     test.comp_het_default.2.db | cols
chrom  start  end    gene    alt  variant_id  comp_het_id  family_id  family_members                                                       family_genotypes         affected_samples  family_count
chr1   17362  17366  WASH7P  T    1_1         1            3          dad_3(dad;unaffected),mom_3(mom;unaffected),child_3(child;affected)  TTCT|T,TTCT|TTCT,TTCT|T  child_3           2
chr1   17362  17366  WASH7P  T    1_1         1            3          dad_3(dad;unaffected),mom_3(mom;unaffected),child_3(child;affected)  TTCT|T,TTCT|TTCT,TTCT|T  child_3           2
chr1   17729  17730  WASH7P  A    2_1         1            3          dad_3(dad;unaffected),mom_3(mom;unaffected),child_3(child;affected)  C|A,C|A,A|C              child_3           2
chr1   17729  17730  WASH7P  A    2_1         1            3          dad_3(dad;unaffected),mom_3(mom;unaffected),child_3(child;affected)  C|A,C|A,A|C              child_3           2
```

**Comment by jxchong** [comment]
I think this looks pretty good.  


On April 14, 2015 at 12:47:19 PM, Brent Pedersen - Bioinformatics (notifications@github.com) wrote:

Here is the output from a few tools. I believe it matches what was proposed. Please let me know any changes:

mendel

brentp@x1:/usr/local/src/gemini/test$ gemini mendel_errors  --columns "chrom, start, end, ref, alt, impact, impact_severity" test.auto_dom.no_parents.db | cols
WARNING: Unable to find parents for family (2). GEMINI is currently only able to identify candidates from two generational families.
chrom  start     end       ref  alt  impact          impact_severity  violation  variant_id         family_id  family_members                                                     family_genotypes  affected_samples  family_count
chr10  1142207   1142208   T    C    stop_loss       HIGH             1          plausible de novo  1          1_dad(dad;unaffected),1_mom(mom;unaffected),1_kid(child;affected)  T/T,T/T,T/C       1_kid             1
chr10  48003991  48003992  C    T    non_syn_coding  MED              3          plausible de novo  1          1_dad(dad;unaffected),1_mom(mom;unaffected),1_kid(child;affected)  C/C,C/C,C/T       1_kid             1
chr10  48004991  48004992  C    T    non_syn_coding  MED              4          plausible de novo  1          1_dad(dad;unaffected),1_mom(mom;unaffected),1_kid(child;affected)  C/C,C/C,C/T       1_kid             1

autosomal dominant

brentp@x1:/usr/local/src/gemini/test$ gemini autosomal_dominant  --columns "chrom, start, end, ref, alt, impact_severity" test.auto_dom.db | cols
chrom  start      end        ref  alt  impact_severity  gene    variant_id  family_id  family_members                                                   family_genotypes  affected_samples  family_count
chr10  48003991   48003992   C    T    MED              ASAH2C  3           3          3_dad(dad;affected),3_mom(mom;unknown),3_kid(child;affected)     C/T,C/C,C/T       3_dad,3_kid       2
chr10  48003991   48003992   C    T    MED              ASAH2C  3           2          2_dad(dad;unaffected),2_mom(mom;affected),2_kid(child;affected)  C/C,C/T,C/T       2_mom,2_kid       2
chr10  48004991   48004992   C    T    MED              ASAH2C  4           3          3_dad(dad;affected),3_mom(mom;unknown),3_kid(child;affected)     C/T,C/C,C/T       3_dad,3_kid       2
chr10  48004991   48004992   C    T    MED              ASAH2C  4           2          2_dad(dad;unaffected),2_mom(mom;affected),2_kid(child;affected)  C/C,C/T,C/T       2_mom,2_kid       2
chr10  135336655  135336656  G    A    LOW              SPRN    5           3          3_dad(dad;affected),3_mom(mom;unknown),3_kid(child;affected)     G/A,G/G,G/A       3_dad,3_kid       1
chr10  1142207    1142208    T    C    HIGH             WDR37   1           3          3_dad(dad;affected),3_mom(mom;unknown),3_kid(child;affected)     T/C,T/T,T/C       3_dad,3_kid       2
chr10  1142207    1142208    T    C    HIGH             WDR37   1           2          2_dad(dad;unaffected),2_mom(mom;affected),2_kid(child;affected)  T/T,T/C,T/C       2_mom,2_kid       2

comp_het

brentp@x1:/usr/local/src/gemini/test$ gemini comp_hets     --ignore-phasing     --columns "chrom, start, end"     --families 3     test.comp_het_default.2.db | cols
chrom  start  end    gene    alt  variant_id  comp_het_id  family_id  family_members                                                       family_genotypes         affected_samples  family_count
chr1   17362  17366  WASH7P  T    1_1         1            3          dad_3(dad;unaffected),mom_3(mom;unaffected),child_3(child;affected)  TTCT|T,TTCT|TTCT,TTCT|T  child_3           2
chr1   17362  17366  WASH7P  T    1_1         1            3          dad_3(dad;unaffected),mom_3(mom;unaffected),child_3(child;affected)  TTCT|T,TTCT|TTCT,TTCT|T  child_3           2
chr1   17729  17730  WASH7P  A    2_1         1            3          dad_3(dad;unaffected),mom_3(mom;unaffected),child_3(child;affected)  C|A,C|A,A|C              child_3           2
chr1   17729  17730  WASH7P  A    2_1         1            3          dad_3(dad;unaffected),mom_3(mom;unaffected),child_3(child;affected)  C|A,C|A,A|C              child_3           2

**Comment by brentp** [comment]
@arq5x and I just chatted. We'll have the standard columns as before and any tool that has additional columns like `comp_het` and `mendel_error` will append the the column at the end. This will simplify using the output from mutliple tools unchanged.

**Comment by arq5x** [comment]
I agree.  @brentp, as we just discussed, I think the only thing to fix is that `comp_hets` has an extra `comp_het_id` column that disrupts the symmetry of the output across the tools.  Let's instead move that column to the end of the output so as to no disrupt the symmetry. Secondly, since `comp_het_id` is its own column, let's just make `variant_id` be the `comp_het_id` from the database, rather than a composite of `variant_id` and `comp_het_id`.

**Comment by arq5x** [comment]
oops, you beat me to it.
